odilia/src/tower.rs

```diff
@@ -24,11 +36,231 @@ use tower::ServiceExt;
 
 use tokio::sync::mpsc::{Receiver, Sender};
 
+use odilia_cache::{AccessiblePrimitive, Cache, CacheItem};
 use odilia_common::command::{
 	CommandType, CommandTypeDynamic, IntoCommands, OdiliaCommand as Command,
-	OdiliaCommandDiscriminants as CommandDiscriminants, Speak,
+	OdiliaCommandDiscriminants as CommandDiscriminants, Speak, TryIntoCommands,
 };
 
+#[derive(Debug, Clone)]
+pub struct CacheEvent<E: EventProperties + Debug> {
+	inner: E,
+	item: CacheItem,
+}
+impl<E> CacheEvent<E>
+where
+	E: EventProperties + Debug,
+{
+	async fn from_event(ev: E, cache: Arc<Cache>) -> Result<Self, Error> {
+		let item = cache.from_event(&ev).await?;
+		Ok(CacheEvent { inner: ev, item })
+	}
+}
+
+#[derive(Clone)]
+pub struct CacheLayer<I> {
+	cache: Arc<Cache>,
+	_marker: PhantomData<fn(I)>,
+}
+impl<I> CacheLayer<I> {
+	fn new(cache: Arc<Cache>) -> Self {
+		CacheLayer { cache, _marker: PhantomData }
+	}
+}
+impl<S, I> Layer<S> for CacheLayer<I>
+where
+	S: Service<(I, Arc<Cache>)>,
+{
+	type Service = CacheService<S, I>;
+	fn layer(&self, inner: S) -> CacheService<S, I> {
+		CacheService { inner, cache: Arc::clone(&self.cache), _marker: PhantomData }
+	}
+}
+pub struct CacheService<S, I> {
```

Comment:
> isn't there already a layer for services which compose with one another?

Tower does provide some types for Layers, but I don't see anything that composes so easily as our own types. Even if we could create them with `layer_fn`, I'd find it easier to debug if we use own names for it instead of trying to debug types that are `LayerFn<...closure...>`. Hopefully that makes sense.

Reply:
> did other crates then have to make their own types as well?

For custom functionality, yes.

Reply:
I recommend using layer_fn and such, but O well, this code is being written in such a way as to hopefully not be prodded in a while, so you're probably be right. And also, that's why we have logs, to not have to rely only on backtraces